File: sdk/python/src/orgo_clone/computer.py
from __future__ import annotations
import base64
from typing import Any
from .client import OrgoClient
from .types import ExecResult, ComputerStatus


class Computer:
    """Represents a virtual desktop computer controlled by AI agents."""

    def __init__(self, client: OrgoClient, data: dict[str, Any]) -> None:
        self._client = client
        self.id: str = data["id"]
        self.name: str = data.get("name", "")
        self.status: ComputerStatus = data.get("status", "creating")
        self.specs: dict[str, Any] = data.get("specs", {})
        self._vnc_url: str | None = data.get("vncUrl") or data.get("vnc_url")
        self._novnc_url: str | None = data.get("novncUrl") or data.get("novnc_url")

    @classmethod
    async def create(
        cls,
        api_key: str,
        workspace_id: str | None = None,
        name: str | None = None,
        cpu: int = 2,
        ram: int = 4096,
        resolution: str = "1280x720x24",
        base_url: str = "http://localhost:3000",
    ) -> "Computer":
        client = OrgoClient(api_key=api_key, base_url=base_url)

        # Auto-create workspace if none specified
        if not workspace_id:
            ws = await client.post("/workspaces", json={"name": name or "default"})
            workspace_id = ws["workspace"]["id"]

        data = await client.post(
            f"/workspaces/{workspace_id}/computers",
            json={"name": name or "computer", "cpu": cpu, "ram": ram, "resolution": resolution},
        )
        return cls(client, data["computer"])

    async def screenshot(self, format: str = "png", quality: int = 70) -> bytes:
        """Take a screenshot and return raw image bytes."""
        data = await self._client.get(
            f"/computers/{self.id}/screenshot",
            params={"format": format, "quality": quality},
        )
        return base64.b64decode(data["image"])

    async def screenshot_base64(self, format: str = "png", quality: int = 70) -> str:
        """Take a screenshot and return base64 string."""
        data = await self._client.get(
            f"/computers/{self.id}/screenshot",
            params={"format": format, "quality": quality},
        )
        return data["image"]

    async def click(self, x: int, y: int) -> None:
        await self._client.post(f"/computers/{self.id}/click", json={"x": x, "y": y})

    async def right_click(self, x: int, y: int) -> None:
        await self._client.post(f"/computers/{self.id}/right-click", json={"x": x, "y": y})

    async def double_click(self, x: int, y: int) -> None:
        await self._client.post(f"/computers/{self.id}/double-click", json={"x": x, "y": y})

    async def scroll(self, x: int = 0, y: int = 0, direction: str = "down", amount: int = 3) -> None:
        await self._client.post(f"/computers/{self.id}/scroll", json={"x": x, "y": y, "direction": direction, "amount": amount})

    async def drag(self, start_x: int, start_y: int, end_x: int, end_y: int) -> None:
        await self._client.post(f"/computers/{self.id}/drag", json={"start_x": start_x, "start_y": start_y, "end_x": end_x, "end_y": end_y})

    async def move(self, x: int, y: int) -> None:
        await self._client.post(f"/computers/{self.id}/move", json={"x": x, "y": y})

    async def type(self, text: str) -> None:
        await self._client.post(f"/computers/{self.id}/type", json={"text": text})

    async def key(self, combo: str) -> None:
        await self._client.post(f"/computers/{self.id}/key", json={"key": combo})

    async def bash(self, command: str, timeout: int = 30) -> ExecResult:
        data = await self._client.post(f"/computers/{self.id}/bash", json={"command": command, "timeout": timeout})
        return data.get("data", data)

    async def exec(self, code: str, language: str = "python") -> ExecResult:
        data = await self._client.post(f"/computers/{self.id}/exec", json={"code": code, "language": language})
        return data.get("data", data)

    async def clipboard_read(self) -> str:
        data = await self._client.get(f"/computers/{self.id}/clipboard")
        return data.get("data", {}).get("text", "")

    async def clipboard_write(self, text: str) -> None:
        await self._client.post(f"/computers/{self.id}/clipboard", json={"text": text})

    async def actions(self, actions: list[dict[str, Any]]) -> list[dict[str, Any]]:
        data = await self._client.post(f"/computers/{self.id}/actions", json={"actions": actions})
        return data.get("results", [])

    async def refresh(self) -> None:
        data = await self._client.get(f"/computers/{self.id}")
        computer = data["computer"]
        self.status = computer.get("status", self.status)
        self.specs = computer.get("specs", self.specs)
        self._vnc_url = computer.get("vncUrl") or computer.get("vnc_url")
        self._novnc_url = computer.get("novncUrl") or computer.get("novnc_url")

    async def stop(self) -> None:
        await self._client.post(f"/computers/{self.id}/stop")
        self.status = "stopping"

    async def start(self) -> None:
        await self._client.post(f"/computers/{self.id}/start")
        self.status = "starting"

    async def destroy(self) -> None:
        await self._client.delete(f"/computers/{self.id}")

    @property
    def vnc_url(self) -> str | None:
        return self._vnc_url if self.status == "running" else None

    @property
    def novnc_url(self) -> str | None:
        return self._novnc_url if self.status == "running" else None

File: sdk/typescript/src/computer.ts
import { OrgoClient } from './client.js';
import type { ComputerStatus, ComputerSpec, ExecResult, CreateComputerOpts, ScreenshotOpts, ScrollOpts, Action, ActionResult } from './types.js';

export class Computer {
  readonly id: string;
  status: ComputerStatus;
  specs: ComputerSpec;
  private _client: OrgoClient;
  private _vncUrl: string | null;
  private _novncUrl: string | null;

  constructor(client: OrgoClient, data: Record<string, unknown>) {
    this._client = client;
    this.id = data.id as string;
    this.status = (data.status as ComputerStatus) || 'creating';
    this.specs = (data.specs as ComputerSpec) || {};
    this._vncUrl = (data.vncUrl ?? data.vnc_url ?? null) as string | null;
    this._novncUrl = (data.novncUrl ?? data.novnc_url ?? null) as string | null;
  }

  static async create(opts: CreateComputerOpts): Promise<Computer> {
    const client = new OrgoClient({ apiKey: opts.apiKey, baseUrl: opts.baseUrl });

    let workspaceId = opts.workspaceId;
    if (!workspaceId) {
      const ws = await client.post<{ workspace: { id: string } }>('/workspaces', { name: opts.name || 'default' });
      workspaceId = ws.workspace.id;
    }

    const data = await client.post<{ computer: Record<string, unknown> }>(
      `/workspaces/${workspaceId}/computers`,
      { name: opts.name || 'computer', cpu: opts.cpu || 2, ram: opts.ram || 4096, resolution: opts.resolution || '1280x720x24' },
    );

    return new Computer(client, data.computer);
  }

  async screenshot(opts?: ScreenshotOpts): Promise<Buffer> {
    const b64 = await this.screenshotBase64(opts);
    return Buffer.from(b64, 'base64');
  }

  async screenshotBase64(opts?: ScreenshotOpts): Promise<string> {
    const params = new URLSearchParams();
    if (opts?.format) params.set('format', opts.format);
    if (opts?.quality) params.set('quality', String(opts.quality));
    const qs = params.toString() ? `?${params}` : '';
    const data = await this._client.get<{ image: string }>(`/computers/${this.id}/screenshot${qs}`);
    return data.image;
  }

  async click(x: number, y: number): Promise<void> {
    await this._client.post(`/computers/${this.id}/click`, { x, y });
  }

  async rightClick(x: number, y: number): Promise<void> {
    await this._client.post(`/computers/${this.id}/right-click`, { x, y });
  }

  async doubleClick(x: number, y: number): Promise<void> {
    await this._client.post(`/computers/${this.id}/double-click`, { x, y });
  }

  async scroll(x: number, y: number, opts?: ScrollOpts): Promise<void> {
    await this._client.post(`/computers/${this.id}/scroll`, { x, y, ...opts });
  }

  async drag(startX: number, startY: number, endX: number, endY: number): Promise<void> {
    await this._client.post(`/computers/${this.id}/drag`, { start_x: startX, start_y: startY, end_x: endX, end_y: endY });
  }

  async move(x: number, y: number): Promise<void> {
    await this._client.post(`/computers/${this.id}/move`, { x, y });
  }

  async type(text: string): Promise<void> {
    await this._client.post(`/computers/${this.id}/type`, { text });
  }

  async key(combo: string): Promise<void> {
    await this._client.post(`/computers/${this.id}/key`, { key: combo });
  }

  async bash(command: string, opts?: { timeout?: number }): Promise<ExecResult> {
    const data = await this._client.post<Record<string, unknown>>(`/computers/${this.id}/bash`, { command, timeout: opts?.timeout });
    const d = (data.data || data) as Record<string, unknown>;
    return { stdout: d.stdout as string || '', stderr: d.stderr as string || '', exitCode: d.exit_code as number || 0, durationMs: d.duration_ms as number || 0 };
  }

  async exec(code: string, opts?: { language?: string }): Promise<ExecResult> {
    const data = await this._client.post<Record<string, unknown>>(`/computers/${this.id}/exec`, { code, language: opts?.language });
    const d = (data.data || data) as Record<string, unknown>;
    return { stdout: d.stdout as string || '', stderr: d.stderr as string || '', exitCode: d.exit_code as number || 0, durationMs: d.duration_ms as number || 0 };
  }

  async clipboardRead(): Promise<string> {
    const data = await this._client.get<{ data: { text: string } }>(`/computers/${this.id}/clipboard`);
    return data.data?.text || '';
  }

  async clipboardWrite(text: string): Promise<void> {
    await this._client.post(`/computers/${this.id}/clipboard`, { text });
  }

  async actions(actions: Action[]): Promise<ActionResult[]> {
    const data = await this._client.post<{ results: ActionResult[] }>(`/computers/${this.id}/actions`, { actions });
    return data.results;
  }

  async refresh(): Promise<void> {
    const data = await this._client.get<{ computer: Record<string, unknown> }>(`/computers/${this.id}`);
    this.status = data.computer.status as ComputerStatus;
    this.specs = (data.computer.specs as ComputerSpec) || {};
  }

  async stop(): Promise<void> {
    await this._client.post(`/computers/${this.id}/stop`);
    this.status = 'stopping';
  }

  async start(): Promise<void> {
    await this._client.post(`/computers/${this.id}/start`);
    this.status = 'starting';
  }

  async destroy(): Promise<void> {
    await this._client.delete(`/computers/${this.id}`);
  }

  get vncUrl(): string | null { return this.status === 'running' ? this._vncUrl : null; }
  get novncUrl(): string | null { return this.status === 'running' ? this._novncUrl : null; }
}
